File: src/format/wal.js
// LevelDB write-ahead-log (.log) parser.
// Block = 32768 bytes. Record header = crc(4) + length(2 LE) + type(1).
//   type: 1=FULL, 2=FIRST, 3=MIDDLE, 4=LAST  (0 = zero/padding)
// A reassembled record is a WriteBatch:
//   header: sequence(8 LE) + count(4 LE); then `count` ops:
//     op: 1=kTypeValue -> key(varlen)+value(varlen); 0=kTypeDeletion -> key(varlen)
import fs from 'node:fs'
import { crc32c, unmaskCrc } from './sstable.js'

const BLOCK = 32768
const HEADER = 7

function readVarint(buf, off) {
  let v = 0, shift = 0, pos = off
  while (true) { const c = buf[pos++]; v += (c & 0x7f) * 2 ** shift; if (!(c & 0x80)) break; shift += 7 }
  return [v, pos]
}

// Return array of {sequence, ops:[{type,key,value}]}
export function parseLog(path) {
  const data = fs.readFileSync(path)
  const batches = []
  let pos = 0
  let frag = null // {parts:[]}

  while (pos + HEADER <= data.length) {
    const blockOff = pos % BLOCK
    if (BLOCK - blockOff < HEADER) { pos += BLOCK - blockOff; continue } // trailer padding
    const storedCrc = data.readUInt32LE(pos)
    const len = data.readUInt16LE(pos + 4)
    const type = data[pos + 6]
    const dataStart = pos + HEADER
    if (type === 0 && len === 0) { // zero padding to block end
      pos += BLOCK - blockOff
      continue
    }
    // A copy taken mid-append can have a torn tail. Verify the record CRC (covers type+data)
    // and treat the first bad/short record as end-of-log (leveldb's own recovery semantics).
    // NOTE: a torn tail here is deliberately NOT "lossy" — the WAL is append-only and we read
    // sequentially from offset 0, so the parsed prefix is a consistent earlier point-in-time and
    // never regresses below an already-read seq; the truncated batch is picked up next refresh.
    if (dataStart + len > data.length) break
    if (unmaskCrc(storedCrc) !== crc32c(data, pos + 6, pos + 7 + len)) break
    const chunk = data.subarray(dataStart, dataStart + len)
    pos = dataStart + len

    if (type === 1) { emit(chunk) }               // FULL
    else if (type === 2) { frag = [chunk] }         // FIRST
    else if (type === 3) { if (frag) frag.push(chunk) } // MIDDLE
    else if (type === 4) { if (frag) { frag.push(chunk); emit(Buffer.concat(frag)); frag = null } } // LAST
  }

  function emit(record) {
    if (record.length < 12) return
    const sequence = Number(record.readBigUInt64LE(0))
    const count = record.readUInt32LE(8)
    let p = 12
    const ops = []
    for (let i = 0; i < count && p < record.length; i++) {
      const opType = record[p++]
      let klen, vlen, key, value
      ;[klen, p] = readVarint(record, p)
      key = record.subarray(p, p + klen); p += klen
      if (opType === 1) { [vlen, p] = readVarint(record, p); value = record.subarray(p, p + vlen); p += vlen }
      ops.push({ type: opType, key: Buffer.from(key), value: value ? Buffer.from(value) : null })
    }
    batches.push({ sequence, ops })
  }

  return batches
}

if (import.meta.url === `file://${process.argv[1].replace(/\\/g, '/')}`) {
  const batches = parseLog(process.argv[2])
  let ops = 0; for (const b of batches) ops += b.ops.length
  console.log(`batches: ${batches.length}, total ops: ${ops}`)
}

File: packages/libzaungast-native/src/format/wal.rs
//! LevelDB write-ahead-log (.log) parser — faithful port of format/chromium/write-ahead-log.ts.
//! Block = 32768 bytes. Record header = crc(4) + length(2 LE) + type(1); type 1=FULL/2=FIRST/
//! 3=MIDDLE/4=LAST. A reassembled record is a WriteBatch: sequence(8 LE) + count(4 LE), then `count`
//! ops (1=put key+value, 0=delete key). A torn tail (bad/short CRC) ends the log (leveldb recovery
//! semantics) — deliberately NOT lossy: the append-only prefix is a consistent earlier point-in-time.

use crate::sstable::{crc32c, unmask_crc};

const BLOCK: usize = 32768;
const HEADER: usize = 7;

pub struct WalOp {
    pub op_type: u8,
    pub key: Vec<u8>,
    pub value: Option<Vec<u8>>,
}
pub struct WalBatch {
    pub sequence: u64,
    pub ops: Vec<WalOp>,
}

fn read_varint(buf: &[u8], off: usize) -> Option<(u64, usize)> {
    let mut v: u64 = 0;
    let mut shift: u32 = 0;
    let mut pos = off;
    loop {
        let c = *buf.get(pos)?;
        pos += 1;
        v |= ((c & 0x7f) as u64) << shift;
        if c & 0x80 == 0 {
            break;
        }
        shift += 7;
        if shift >= 64 {
            return None;
        }
    }
    Some((v, pos))
}

fn emit(record: &[u8], batches: &mut Vec<WalBatch>) {
    if record.len() < 12 {
        return;
    }
    let sequence = u64::from_le_bytes(record[0..8].try_into().unwrap());
    let count = u32::from_le_bytes(record[8..12].try_into().unwrap());
    let mut p = 12usize;
    let mut ops = Vec::new();
    let mut i = 0u32;
    while i < count && p < record.len() {
        let op_type = record[p];
        p += 1;
        let Some((klen, np)) = read_varint(record, p) else {
            break;
        };
        p = np;
        let klen = klen as usize;
        if p + klen > record.len() {
            break;
        }
        let key = record[p..p + klen].to_vec();
        p += klen;
        let value = if op_type == 1 {
            let Some((vlen, np)) = read_varint(record, p) else {
                break;
            };
            p = np;
            let vlen = vlen as usize;
            if p + vlen > record.len() {
                break;
            }
            let v = record[p..p + vlen].to_vec();
            p += vlen;
            Some(v)
        } else {
            None
        };
        ops.push(WalOp {
            op_type,
            key,
            value,
        });
        i += 1;
    }
    batches.push(WalBatch { sequence, ops });
}

pub fn parse_write_ahead_log(path: &str) -> std::io::Result<Vec<WalBatch>> {
    let data = std::fs::read(path)?;
    let mut batches: Vec<WalBatch> = Vec::new();
    let mut pos = 0usize;
    let mut frag: Option<Vec<u8>> = None;

    while pos + HEADER <= data.len() {
        let block_off = pos % BLOCK;
        if BLOCK - block_off < HEADER {
            pos += BLOCK - block_off; // trailer padding
            continue;
        }
        let stored_crc = u32::from_le_bytes(data[pos..pos + 4].try_into().unwrap());
        let len = u16::from_le_bytes(data[pos + 4..pos + 6].try_into().unwrap()) as usize;
        let rtype = data[pos + 6];
        let data_start = pos + HEADER;
        if rtype == 0 && len == 0 {
            pos += BLOCK - block_off; // zero padding to block end
            continue;
        }
        if data_start + len > data.len() {
            break; // torn tail → end of log
        }
        if unmask_crc(stored_crc) != crc32c(&data, pos + 6, pos + 7 + len) {
            break; // bad crc → end of log
        }
        let chunk = &data[data_start..data_start + len];
        pos = data_start + len;

        match rtype {
            1 => emit(chunk, &mut batches),   // FULL
            2 => frag = Some(chunk.to_vec()), // FIRST
            3 => {
                if let Some(f) = frag.as_mut() {
                    f.extend_from_slice(chunk);
                }
            } // MIDDLE
            4 => {
                if let Some(mut f) = frag.take() {
                    f.extend_from_slice(chunk);
                    emit(&f, &mut batches);
                }
            } // LAST
            _ => {}
        }
    }
    Ok(batches)
}
